spec/controllers/budgets_controller_spec.rb

```diff
@@ -0,0 +1,87 @@
+require "rails_helper"
+
+describe BudgetsController do
+  context "GET #new" do
+    context "when invalid params" do
+      it "won't create a new budget" do
+        user = create(:user)
+        allow(controller).to receive(:current_user).and_return(user)
+        budget_params = { budget_type: "Holiday" }
+
+        expect do
+          post :create, params: { budget: budget_params }
+        end.to change(Budget, :count).by(0)
+      end
+
+      it "renders new" do
+        user = create(:user)
+        allow(controller).to receive(:current_user).and_return(user)
+        budget_params = { budget_type: "Holiday" }
+
+        post :create, params: { budget: budget_params }
+
+        expect(response).to render_template(:new)
+      end
+
+      it "sets the flash[:alert]" do
+        user = create(:user)
+        allow(controller).to receive(:current_user).and_return(user)
+        budget_params = { budget_type: "Holiday" }
+
+        post :create, params: { budget: budget_params }
+
+        expect(flash[:alert]).
+          to match I18n.t("flash.actions.create.alert", resource_name: "Budget")
+      end
+    end
+  end
+
+  context "GET #edit" do
+    context "when invalid parameters" do
+      it "won't update the budget attributes" do
+        user = create(:user)
+        allow(controller).to receive(:current_user).and_return(user)
+        budget = create(:budget, name: "Holiday")
```

Comment:
I would write this spec as follows:
```rb
user =  create(:user)
budget = create(:budget, name: "Holiday", user: user)
allow(controller).to receive(:current_user).and_return(user)
```

This way, there is no confusion if this budget belongs to that user.

Reply:
There should also be a spec that a user cannot edit someone else's budgets

Reply:
That looks great thanks, done in sha: 2823794